quantum/impl/quantum_read_write_mutex_impl.h

```diff
@@ -143,35 +175,40 @@ inline
 ReadWriteMutex::ReadGuard::ReadGuard(ReadWriteMutex& lock,
                                      ReadWriteMutex::TryToLock) :
     _mutex(&lock),
-    _ownsLock(_mutex->tryLockRead())
+    _ownsLock(_mutex->tryLockRead()),
+    _isUpgraded(false)
 {
 }
 
 inline
 ReadWriteMutex::ReadGuard::ReadGuard(ReadWriteMutex& lock,
                                      ReadWriteMutex::AdoptLock) :
     _mutex(&lock),
-    _ownsLock(true)
+    _ownsLock(lock.isReadLocked()),
+    _isUpgraded(lock.isWriteLocked())
```

Comment:
Similarly to Mutex::Guard, we should look at adding APIs to ReadWriteMutex to identify if the caller owns the lock, and assert based on those APIs for the guard AdoptLock constructors. This has to wait for the fully owner-aware ReadWriteMutex.